README.md
# Small_tools
Small tools used to manipulate file formats and process small amounts of data.


CHANGELOG.md

0.1 (2018-08-01)
----------------
Initial release:

- Template deloyment scripts
- filter_duplicated_unmapped.py: Removal of duplicate unmapped BAM file reads
- github_backup.py: Local archiving of Edinburgh Genomics GitHub repositories



tests/test_convert_genotyping_to_vcf.py
import os
from unittest.case import TestCase
from convert_genotyping_to_vcf import convert_genotype_csv

__author__ = 'tcezard'


class Test_small_tools(TestCase):
    def setUp(self):
        self.genotype_csv = os.path.join(os.path.dirname(__file__), 'test_data','E03159_WGS_32_panel_9504430.csv')
        self.genotype_csv = os.path.join(os.path.dirname(__file__), 'test_data','E03159_WGS_32_panel.csv')
        self.small_reference_fai = os.path.join(os.path.dirname(__file__), 'test_data','genotype_32_SNPs_genome_600bp.fa.fai')
        self.reference_fai = os.path.join(os.path.dirname(__file__), 'test_data','GRCh37.fa.fai')



    def test_convert_genotype_csv(self):
        vcf_text = """#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	50293
1	163289571	rs2136241	C	T	.	.	GT	1/1
1	208068579	rs2259397	T	A	.	.	GT	1/1
1	59569829	rs3010325	C	T	.	.	GT	0/1
10	117923225	rs4751955	A	.	.	.	GT	./.
10	1511786	rs1533486	T	G	.	.	GT	0/1
12	23769449	rs10771010	T	C	.	.	GT	1/1
12	28781965	rs12318959	C	.	.	.	GT	./.
13	43173198	rs3742257	T	.	.	.	GT	0/0
14	25843774	rs1377935	T	.	.	.	GT	0/0
14	55932919	rs946065	C	.	.	.	GT	./.
15	99130113	rs6598531	T	.	.	.	GT	0/0
16	82622140	rs4783229	T	C	.	.	GT	0/1
18	35839365	rs1567612	G	A	.	.	GT	0/1
18	42481985	rs11660213	A	G	.	.	GT	0/1
19	39697974	rs11083515	A	G	.	.	GT	1/1
2	11200347	rs7564899	G	A	.	.	GT	0/1
2	21084332	rs4971536	C	.	.	.	GT	0/0
2	50525067	rs10194978	G	A	.	.	GT	0/1
3	181638250	rs4855056	A	G	.	.	GT	0/1
5	11870138	rs6554653	C	T	.	.	GT	1/1
6	125039942	rs1415762	C	.	.	.	GT	0/0
6	163719115	rs6927758	C	.	.	.	GT	./.
6	25548288	rs441460	G	A	.	.	GT	1/1
6	37572144	rs7773994	T	.	.	.	GT	0/0
6	9914294	rs9396715	T	.	.	.	GT	0/0
7	126113335	rs7796391	A	G	.	.	GT	1/1
8	1033625	rs2336695	A	G	.	.	GT	1/1
8	104215466	rs1157213	T	C	.	.	GT	0/1
9	80293657	rs10869955	C	A	.	.	GT	0/1
Y	14850341	rs2032598	T	.	.	.	GT	./.
Y	6818291	rs768983	C	.	.	.	GT	./.
Y	8602518	rs3913290	C	.	.	.	GT	./."""
        #text=convert_genotype_csv(self.genotype_csv,self.reference_fai)
        #print(text)
        #self.assertEqual(text,vcf_text)

        text=convert_genotype_csv(self.genotype_csv, self.small_reference_fai, 600)

        print(text)

tests/test_filter_duplicated_unmapped.py
import os
from unittest.case import TestCase
from unittest.mock import patch, call
from python_scripts.filter_duplicated_unmapped import filter_duplicated_unmpped_read


class TestFilterDupUnmapped(TestCase):
    asset_dir = os.path.join(os.path.dirname(__file__), 'assets')

    def test_filter_duplicated_unmpped_read(self):
        input_file = os.path.join(self.asset_dir, 'test.bam')
        output_file = os.path.join(self.asset_dir, 'test_out.bam')
        with patch('sys.stderr.write') as patch_write:
            filter_duplicated_unmpped_read(input_file, output_file)
            assert patch_write.mock_calls[0] == call('Read 17 reads from %s\n' % input_file)
            assert patch_write.mock_calls[1] == call('Skipped 1 reads\n')

    def tearDown(self):
        output_file = os.path.join(self.asset_dir, 'test_out.bam')
        if os.path.exists(output_file):
            os.unlink(output_file)
            os.unlink(output_file + '.bai')


python_scripts/github_backup.py
import os
import requests
import argparse
from datetime import datetime
from time import sleep


class BackupError(Exception):
    pass


def request(method, endpoint, **kwargs):
    response = session.request(method, 'https://api.github.com' + endpoint, **kwargs)
    if not 200 <= response.status_code <= 299:
        raise BackupError('%s on %s got status %s: %s' % (method, endpoint, response.status_code, response.content))

    return response

session = requests.Session()
session.headers['Accept'] = 'application/vnd.github.wyandotte-preview+json'
repos = [
    'EdinburghGenomics/' + r
    for r in ('Analysis-Driver', 'Reporting-App', 'EGCG-Project-Management', 'EGCG-Core', 'Fastq-Filterer')
]


def main():
    a = argparse.ArgumentParser()
    a.add_argument(
        'key_file',
        help="""Path to a single-line file containing GitHub API credentials in the format
                'username:personal_access_token'. The access token requires the scope 'admin:org'."""
    )
    a.add_argument('--dest', help='Directory to save archives to (defaults to working dir)', default=os.getcwd())
    a.add_argument('--quiet', dest='stdout', action='store_false')
    args = a.parse_args()

    if args.stdout:
        log = print
    else:
        def log(msg):
            pass

    with open(args.key_file, 'r') as f:
        username, token = f.readline().rstrip('\n').split(':')
        session.auth = requests.auth.HTTPBasicAuth(username, token)

    log('Initiating migration')
    migration_id = request('POST', '/orgs/EdinburghGenomics/migrations', json={'repositories': repos}).json()['id']
    log('Waiting for new migration %s to finish' % migration_id)
    while True:
        sleep(10)
        status = request('GET', '/orgs/EdinburghGenomics/migrations/%s' % migration_id).json()['state']
        if status == 'exported':
            break
        elif status in ('exporting', 'pending'):
            pass
        else:
            raise BackupError('Got status %s' % status)

    log('Downloading migration archive')
    archive = request('GET', '/orgs/EdinburghGenomics/migrations/%s/archive' % migration_id)
    tar_file = 'EdinburghGenomics_archive_%s.tar.gz' % datetime.now().strftime('%Y-%m-%d_%H:%M:%S')
    with open(os.path.join(args.dest, tar_file), 'wb') as f:
        f.write(archive.content)

    log('Cleaning up server-side archive')
    request('DELETE', '/orgs/EdinburghGenomics/migrations/%s/archive' % migration_id)

    log('Done')


if __name__ == '__main__':
    main()


deployment_scripts/python_project_in_venv.sh
#!/bin/bash
# python_project_in_venv.sh: deployment script for installing a project to a
# Python virtual environment to run as an executable

### config
# pyvenv: path to central executable for installing Python virtual environments
pyvenv=

# repo: remote url of the git repository
repo=

# any preliminary commands, exports, etc

if [ ! $pyvenv ] || [ ! $repo ]
then
    echo "Invalid config: ensure that pyvenv and repo are populated"
    exit 1
fi
### end config


if [ $# == 0 ]
then
    echo "Usage: $0 <git_tag> [setlink]"
    exit
fi

scriptpath=$(dirname $(readlink -f $0))
cd $scriptpath
git_tag=$1


if [ "$git_tag" == "master" ]
then
    deployment_suffix=""
else
    deployment_suffix="@$git_tag"
fi

echo "Deploying $git_tag from $repo"

if [ -d "./$git_tag" ]
then
  rm -rf "./$git_tag"
fi

$pyvenv $git_tag
$git_tag/bin/pip install -q "git+$repo$deployment_suffix"

if [ "$2" == "setlink" ]
then
    echo "Setting production link"
    ln -fnvs ./$git_tag ./production
fi

echo "Done"


python_scripts/filter_duplicated_unmapped.py
#!/usr/bin/env python
import os
import sys
from argparse import ArgumentParser
import pysam


def filter_duplicated_unmpped_read(input_file_path, output_file_path):
    infile = pysam.AlignmentFile(input_file_path, "rb")
    outfile = pysam.AlignmentFile(output_file_path, "wb", template=infile)
    total_reads = reads_skipped = 0
    for s in infile:
        total_reads += 1
        if s.is_unmapped and not s.mate_is_unmapped and s.has_tag('BD'):
            # Skip that read
            reads_skipped += 1
        else:
            outfile.write(s)

    infile.close()
    outfile.close()
    sys.stderr.write('Read %s reads from %s\n' % (total_reads, input_file_path))
    sys.stderr.write('Skipped %s reads\n' % reads_skipped)
    sys.stderr.write('Wrote %s reads to %s\n' % (total_reads - reads_skipped, output_file_path))

    if os.path.exists(output_file_path):
        sys.stderr.write('Indexing %s\n' % output_file_path)
        pysam.index(output_file_path)

    return 0


def main():
    argparser = _prepare_argparser()
    args = argparser.parse_args()
    return filter_duplicated_unmpped_read(args.input_bam_file, args.output_bam_file)


def _prepare_argparser():
    """Prepare optparser object. New arguments will be added in this function first."""
    argparser = ArgumentParser()
    argparser.add_argument("-i", "--input_bam_file", type=str, required=True,
                           help="The bam file containing the reads to filter.")
    argparser.add_argument("-o", "--output_bam_file", type=str, required=True,
                           help="The bam file created after filtering.")
    return argparser


if __name__ == "__main__":
    sys.exit(main())


deployment_scripts/python_project_from_source.sh
#!/bin/bash
# python_project_from_source.sh: deployment script for setting up a project to run
# from source

### config
# pip: relative or absolute path to pip
pip=

# repo: remote url of the git repository
repo=

# any preliminary commands, exports, etc

if [ ! $pip ] || [ ! $repo ]
then
    echo "Invalid config: ensure that pip and repo are populated"
    exit 1
fi
### end config


if [ $# == 0 ]
then
    echo "Usage: $0 <git_tag> [setlink]"
    exit
fi

scriptpath=$(dirname $(readlink -f $0))
cd $scriptpath
git_tag=$1

if [ ! -d git_repo ]
then
    echo "Initialising local Git repo"
    git init --bare --shared git_repo
fi

echo "Deploying $git_tag from $repo"

git --git-dir=git_repo fetch --tags $repo
mkdir $git_tag
git --git-dir=git_repo --work-tree=$git_tag checkout -f $git_tag
chmod -R a-w $git_tag

$pip install -q -r $git_tag/requirements.txt

if [ "$2" == "setlink" ]
then
    echo "Setting production link"
    ln -fnvs $git_tag production
fi

echo "Done"


convert_genotyping_to_vcf.py
from argparse import ArgumentParser
from collections import defaultdict
import csv

__author__ = 'tcezard'

SNPs_definition = {"C___2728408_10": ["rs3010325",  "1",  "59569829",  "C", "T", "Reverse"],
                   "C___1563023_10": ["rs2136241",  "1",  "163289571", "C", "T", "Reverse"],
                   "C__15935210_10": ["rs2259397",  "1",  "208068579", "T", "C", "Reverse"],
                   "C__33211212_10": ["rs7564899",  "2",  "11200347",  "G", "A", "Forward"],
                   "C___3227711_10": ["rs4971536",  "2",  "21084332",  "C", "T", "Reverse"],
                   "C__30044763_10": ["rs10194978", "2",  "50525067",  "G", "A", "Forward"],
                   "C__11821218_10": ["rs4855056",  "3",  "181638250", "A", "G", "Forward"],
                   "C___1670459_10": ["rs6554653",  "5",  "11870138",  "C", "T", "Reverse"],
                   "C__29619553_10": ["rs9396715",  "6",  "9914294",   "T", "C", "Reverse"],
                   "C___1007630_10": ["rs441460",   "6",  "25548288",  "G", "A", "Reverse"],
                   "C__26546714_10": ["rs7773994",  "6",  "37572144",  "T", "G", "Forward"],
                   "C___7421900_10": ["rs1415762",  "6",  "125039942", "C", "T", "Forward"],
                   "C__27402849_10": ["rs6927758",  "6",  "163719115", "C", "T", "Reverse"],
                   "C___2953330_10": ["rs7796391",  "7",  "126113335", "A", "G", "Reverse"],
                   "C__16205730_10": ["rs2336695",  "8",  "1033625",   "A", "G", "Forward"],
                   "C___8850710_10": ["rs1157213",  "8",  "104215466", "T", "C", "Forward"],
                   "C___1801627_20": ["rs10869955", "9",  "80293657",  "C", "A", "Reverse"],
                   "C___7431888_10": ["rs1533486",  "10", "1511786",   "T", "G", "Forward"],
                   "C___1250735_20": ["rs4751955",  "10", "117923225", "A", "G", "Forward"],
                   "C___1902433_10": ["rs10771010", "12", "23769449",  "T", "C", "Forward"],
                   "C__31386842_10": ["rs12318959", "12", "28781965",  "C", "T", "Reverse"],
                   "C__26524789_10": ["rs3742257",  "13", "43173198",  "T", "C", "Forward"],
                   "C___8924366_10": ["rs1377935",  "14", "25843774",  "T", "C", "Reverse"],
                   "C_____43852_10": ["rs946065",   "14", "55932919",  "C", "A", "Forward"],
                   "C__11522992_10": ["rs6598531",  "15", "99130113",  "T", "G", "Forward"],
                   "C__10076371_10": ["rs4783229",  "16", "82622140",  "T", "C", "Reverse"],
                   "C___7457509_10": ["rs1567612",  "18", "35839365",  "G", "A", "Forward"],
                   "C___1122315_10": ["rs11660213", "18", "42481985",  "A", "G", "Reverse"],
                   "C__11710129_10": ["rs11083515", "19", "39697974",  "A", "G", "Forward"],
                   "C___1027548_20": ["rs768983",   "Y",  "6818291",   "C", "T", "Reverse"],
                   "C___8938211_20": ["rs3913290",  "Y",  "8602518",   "C", "T", "Forward"],
                   "C___1083232_10": ["rs2032598",  "Y",  "14850341",  "T", "C", "Reverse"]}

HEADERS_CALL="Call"
#Actual Header in the file
HEADERS_SAMPLE_ID = ["StudyID", 'Sample ID']
HEADERS_ASSAY_ID=["SNPName", "Assay Name"]

vcf_header=['#CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO','FORMAT']

_complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}
def complement(bases):
    return ''.join([_complement.get(base) for base in bases])

def get_genotype_from_call(ref_allele, alternate_allele, call, design_strand):
    """Uses the SNPs definition to convert the genotype call to a vcf compatible genotype"""
    genotype = './.'
    if call == 'Undefined':
        return alternate_allele, genotype
    if call == 'Both':
        call=ref_allele + alternate_allele
    #The call arrive on the forward strand so no need to reverse it
    #if design_strand == 'Reverse':
    #    call = complement(call)
    callset = set(call)
    if ref_allele in callset and len(callset) == 1:
        genotype='0/0'
    elif ref_allele in callset and len(callset) == 2:
        genotype='0/1'
        callset.remove(ref_allele)
        alternate_allele = ','.join(callset)
    elif not ref_allele in callset and len(callset) == 1:
        genotype='1/1'
        alternate_allele = ','.join(callset)
    return alternate_allele, genotype

def vcf_header_from_fai_file(genome_fai):
    """Generate a vcf header from an fai file"""
    header_entries = []
    with open(genome_fai) as open_file:
        reader = csv.reader(open_file, delimiter='\t')
        for row in reader:
            header_entries.append('##contig=<ID=%s,length=%s>'%(row[0], row[1]))
    return header_entries

def order_from_fai(all_records, genome_fai):
    ordered_records = []
    with open(genome_fai) as open_file:
        reader = csv.reader(open_file, delimiter='\t')
        for row in reader:
            snps = all_records.get(row[0], [])
            #Sort the SNPs by position within a ref
            snps.sort(key=lambda snp: int(snp[1]))
            #convert the array to str
            ordered_records.extend(['\t'.join(s) for s in snps])
    return ordered_records

def convert_genotype_csv(csv_file, genome_fai, flank_length=0):
    all_samples = set()
    with open(csv_file) as open_file:
        reader = csv.DictReader(open_file, delimiter='\t')
        all_records = defaultdict(dict)
        fields = set(reader.fieldnames)
        for h in  HEADERS_SAMPLE_ID:
            if h in fields:
                header_sample_id = h
        for h in  HEADERS_ASSAY_ID:
            if h in fields:
                header_assay_id = h
        for line in reader:
            sample = line[header_sample_id]
            if sample.lower() =='blank':
                #Entries with blank as sample name are entries with water and no DNA
                continue
            assay_id = line[header_assay_id]
            SNPs_id, reference_name, reference_position, ref_allele, alt_allele, design_strand = SNPs_definition.get(assay_id)
            #alt_allele is the alternate allele from the dbsnp definition
            #It will be replaced by the alt allele from the call if it exists
            alt_allele, genotype = get_genotype_from_call(ref_allele, alt_allele, line.get(HEADERS_CALL), design_strand)
            if not 'SNPs' in all_records[SNPs_id]:
                if flank_length:
                    SNP=[assay_id, str(flank_length+1), SNPs_id, ref_allele, alt_allele, ".", ".", ".", "GT"]
                else:
                    SNP=[reference_name, reference_position, SNPs_id, ref_allele, alt_allele, ".", ".", ".", "GT"]
                all_records[SNPs_id]['SNP']=SNP
            if sample in all_records[SNPs_id]:
                raise Exception('Sample {} found more than once for SNPs {}'.format(sample, SNPs_id))
            all_records[SNPs_id][sample]=genotype
            all_samples.add(sample)

    all_samples = sorted(all_samples)
    SNPs_list = defaultdict(list)
    for SNPs_id in  all_records:
        record = all_records.get(SNPs_id)
        out = record.get('SNP')
        out.extend([record.get(sample) for sample in all_samples])
        SNPs_list[out[0]].append(out)

    all_lines = ["##fileformat=VCFv4.1",
                 '##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">']
    all_lines.extend(vcf_header_from_fai_file(genome_fai))
    vcf_header.extend(all_samples)
    all_lines.append('\t'.join(vcf_header))
    all_lines.extend(order_from_fai(SNPs_list, genome_fai))
    return '\n'.join(all_lines)

def main():
    argparser = _prepare_argparser()
    args = argparser.parse_args()
    with open(args.output_file, 'w') as open_ouput:
        text = convert_genotype_csv(args.genotype_file, args.reference_fai, args.flanking_size)
        open_ouput.write(text)

def _prepare_argparser():
    """Prepare optparser object. New arguments will be added in this
    function first.
    """
    description = """Convert genotype file to vcf for comparison"""

    argparser = ArgumentParser(description=description)

    argparser.add_argument("-g", "--genotype_file", dest="genotype_file", type=str, required = True,
                           help="The genotype file to convert.")
    argparser.add_argument("-r", "--reference_fai", dest="reference_fai", type=str, required = True,
                           help="The file containing the reference fai.")
    argparser.add_argument("-f", "--flanking_size", dest="flanking_size", type=int,
                           help="The size of the flanking region used to build the small genome.")
    argparser.add_argument("-o", "--output_file", dest="output_file", type=str, required = True,
                           help="The output file name.")
    return argparser

if __name__=="__main__":
    main()